### DISCUSSION TITLE
Support for using dynamic task mapping with templated arguments on `KubernetesPodOperator`

### DISCUSSION THREAD

**Discussion by orphefs** [apache/airflow#29660] [Q&A]
Dear community,

As part of a pipelining workflow to run containers on k8s, one of the things I am trying to achieve is to dynamically name a PVC (with dynamically provisioned PV ) by generating a hashed string in task `generate_pvc_name()` which I am then using in a downstream task where I am using the k8s python API to create the PVC. 

The idea behind this is to avoid PVC name collisions when running multiple DAGs in parallel. I am passing data between tasks using XComs. The aforementioned works well, until I introduce dynamic task mapping:

In a downstream task  I am introducing a [KubernetesPodOperator](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html) with [dynamic task mapping](https://airflow.apache.org/docs/apache-airflow/2.3.0/concepts/dynamic-task-mapping.html) . There I could traditionally pass a templated argument to `volumes` to initialize the operator, i.e.

```python
task_execute_mbm_docker_container = KubernetesPodOperator.partial(
    volumes=[instantiate_persistent_volume_obj("{{ ti.xcom_pull(task_ids='generate_pvc_name') }}")],
    volume_mounts=[instantiate_volume_mount_obj()],
    name="my_docker_container",
    image="docker-registry.company.com/my_container:latest",
    cmds=["bash"],
    task_id="execute_docker_container",
    do_xcom_push=False,
    dag=dag_dynamic_task_mapping,
    **global_task_args
).expand(arguments=[task_execute_mbm_docker_container_bash_command(filename) for filename in filenames],
         )
```

but it seems that templates are not supported well within dynamic task mapping which means the template is not evaluated, as mentioned [here](https://airflow.apache.org/docs/apache-airflow/2.3.0/concepts/dynamic-task-mapping.html#how-do-templated-fields-and-mapped-arguments-interact) so I was wondering if you would be aware of any workarounds to this. This [link](https://www.mail-archive.com/users@airflow.apache.org/msg00861.html) offers some pointers,but it is not quite exactly what I would like to achieve.

Another idea I had is to generate a pod template file containing the PVC and volumeMount specs, and reuse it throughout all `KubernetesPodOperator`'s that need it. However, if I instantiate this as a global variable in my DAG, I have no control over when and where the expression gets evaluated, and because it will contain a random string, I will end up with a different version of the pod template file at every task. 

I would greatly appreciate any pointers/ideas/guidance on how to achieve this. I am looking forward to your reply.

**Comment by potiuk** [comment]
I think this is simply not possible now in Dynamic Task Mapping case (though I agree it would be great to have it)..

I think excluding the mapped parameters from being templated was a deliberate decision which was connected with optimisations, but there should be no fundamental reasons to implement it similarly to implementing default_args https://github.com/apache/airflow/issues/29366  - it's just "this does not work yet" case.  CC: @uranusjr - I based it on your comments in #29366 and the docs, so you might correct me if I am wrong. Allowing templated fields to be mapped and remain mapped simply adds quite some complexity and likely hampers performance of dynamic task mapping. 

I think a good workaround will be to pre-generate the names of PVC volumes in the task that calls `expand_kwargs` (https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/dynamic-task-mapping.html#assigning-multiple-parameters-to-a-non-taskflow-operator)  and make them part of the mapping. Not sure however whether you can pass k8s.volume this way (but even if not, maybe the legacy "dict" method of pasing the volumes might work).

**Reply by orphefs** [reply]
Thanks @potiuk - I have found a workaround which involves augmenting the `k8s` classes like so 

```python
from kubernetes.client import models as k8s

# this class is augmented to enable template rendering with dynamic task mapping
class PatchedV1Volume(k8s.V1Volume):
    template_fields = ("persistent_volume_claim",)
    task_type = k8s.V1Volume.__name__

# this class is augmented to enable template rendering with dynamic task mapping
class PatchedV1PersistentVolumeClaimVolumeSource(k8s.V1PersistentVolumeClaimVolumeSource):
    template_fields = ("claim_name", )
    task_type = k8s.V1PersistentVolumeClaimVolumeSource.__name__

```

and then I can use the above within `KubernetesPodOperator` like so 

```python
task_execute_docker_container = KubernetesPodOperator.partial(
    volumes=[PatchedV1Volume(name='airflow-test-volume',
                        persistent_volume_claim=PatchedV1PersistentVolumeClaimVolumeSource(
                            claim_name=str(task_generate_pvc_name.output))
                        )],
    volume_mounts=[instantiate_volume_mount_obj()],
    name="docker_container",
    image="example.com/myimage:latest",
    cmds=["bash"],
    task_id="execute_docker_container",
    dag=dag_dynamic_task_mapping,
    **global_task_args
).expand(arguments=[task_execute_docker_container_bash_command(filename) for filename in filenames],
         )
```

The `task_generate_pvc_name` task returns a string via XCom.

This works with `apache-airflow==2.5.0` and `apache-airflow-providers-cncf-kubernetes==5.1.1`.

**Reply by raphaelauv** [reply]
@orphefs 

since 5.1.0 there is templating of  KPO volumes -> https://github.com/apache/airflow/pull/27719

why do you create PatchedV1Volume and PatchedV1PersistentVolumeClaimVolumeSource ?

**Reply by uranusjr** [reply]
Yes, there’s no technical reason this cannot work, but someone needs to implement it.
